pytorch_lightning/callbacks/swa.py

```diff
@@ -0,0 +1,228 @@
+                module.running_mean = torch.zeros_like(
+                    module.running_mean, device=device, dtype=module.running_mean.dtype)
+                module.running_var = torch.ones_like(
+                    module.running_var, device=device, dtype=module.running_var.dtype)
+                self.momenta[module] = module.momentum
+                module.momentum = None
+                module.num_batches_tracked *= 0
+
+    def on_fit_start(self, trainer, pl_module):
+        self._average_model = deepcopy(pl_module).to("cpu")
+        optimizers = trainer.optimizers
+        lr_schedulers = trainer.lr_schedulers
+
+        if len(optimizers) > 1:
+            raise MisconfigurationException("SWA currently not supported for more than 1 optimizer.")
+
+        if len(lr_schedulers) > 1:
+            raise MisconfigurationException("SWA currently not supported for more than 1 lr_scheduler.")
+
+        self._max_epochs = trainer.max_epochs
+
+        # convert float to integer.
+        if isinstance(self._swa_epoch_start, float):
+            self._swa_epoch_start = int(self._max_epochs * self._swa_epoch_start)
+
+        self._model_contains_batch_norm = self.pl_module_contains_batch_norm(pl_module)
+
+        if self._model_contains_batch_norm:
+            # virtually increase max_epochs to perform batch norm update on latest epoch.
+            trainer.max_epochs += 1
+
+    def on_train_epoch_start(self, trainer, pl_module):
+        if trainer.current_epoch == self._swa_epoch_start:
+            optimizers = trainer.optimizers
+            lr_scheduler = trainer.lr_schedulers[0]["scheduler"]
+
+            self._swa_scheduler = SWALR(
+                optimizers[0],
+                swa_lr=self._swa_lrs,
+                anneal_epochs=self._annealing_epochs,
+                anneal_strategy=self._annealing_strategy,
+                last_epoch=trainer.max_epochs if self._annealing_strategy == "cos" else -1
+            )
+
+            rank_zero_warn(f"swapping lr_scheduler {lr_scheduler} for {self._swa_scheduler}")
+
+            trainer.lr_schedulers[0]["scheduler"] = self._swa_scheduler
+
+            self.n_averaged = torch.tensor(0, dtype=torch.long, device=pl_module.device)
+
+        if self._swa_epoch_start <= trainer.current_epoch and trainer.current_epoch < self._max_epochs:
+            self.update_parameters(self._average_model, pl_module, self.n_averaged, self.avg_fn)
+
+        if trainer.current_epoch == self._max_epochs:
+            if not self._model_contains_batch_norm:
+                self.update_parameters(self._average_model, pl_module, self.n_averaged, self.avg_fn)
+
+            # Transfer weights from average model to pl_module
+            self.transfer_weights(self._average_model, pl_module)
+
+            if self._model_contains_batch_norm:
+                # Only update batch norm stats - Freeze the rest
+                BaseFinetuningCallback.freeze(pl_module, train_bn=True)
```

Comment:
same here, this would be always True
```suggestion
            # Only update batch norm stats - Freeze the rest
            BaseFinetuningCallback.freeze(pl_module, train_bn=True)
```

Reply:
The model should be frozen only when it contains  batch_norm for max_epochs + 1, so only the bachnorm gets trained.